T/TVM/build_tarballs.jl

```diff
@@ -60,8 +54,8 @@ products = Product[
 
 # Dependencies that must be installed before this package can be built
 dependencies = [
-    "Zlib_jll",
-    "LLVM_jll",
+    Dependency(PackageSpec(name="Zlib_jll")),
+    Dependency(PackageSpec(name="LLVM_jll")),
```

Comment:
I'm not entirely sure, but I have the feeling llvm should be a build dependency https://juliapackaging.github.io/BinaryBuilder.jl/dev/reference/#BinaryBuilderBase.BuildDependency

Reply:
That could very well be, I just followed BinaryBuilder's string -> Dependency conversion behaviour here. Also, the [TVM installation docs](https://tvm.apache.org/docs/install/from_source.html) state:

> TVM optionally depends on LLVM. LLVM is required for CPU codegen that needs LLVM

Would this pick up on Julia's LLVM, or does it require the JLL to be present at install time?

Reply:
Yes, my point is that I _think_ this should link to Julia's LLVM, but I don't think we have other libraries linking to LLVM, so I'm not sure how this would work

Reply:
@staticfloat @vchuravy can one of you please confirm this is the right thing to do?

Reply:
TVM will need libLLVM at runtime.  I would just try it with a normal `Dependency("LLVM_jll")` and see if it works; I don't know how sensitive it is to different LLVM versions.

Reply:
`LLVM_jll` doesn't even have libllvm among products. `libLLVM_jll` does, but the library isn't dlopened

Reply:
To be clear, I'm more worried about what happens if julia tries to dlopen another linllvm

Reply:
I thought `LLVM_jll` depends on `libLLVM_jll`? Trying with just libLLVM as a (Build)Dependency results in
`Warning: Linked library libLLVM-9.so could not be resolved and could not be auto-mapped` and failed dlopen calls of all the TVM libraries during audit.

The docs mention that TVM can work with LLVM 4+, so presumably all Julia-compatible versions are supported?

Reply:
Reverted to having `LLVM_jll` as a normal dependency for now. I'm a little confused as to what exactly is needed for runtime codegen, as GPUCompiler uses LLVM.jl (which doesn't seem to rely on the JLL) and https://github.com/JuliaPackaging/Yggdrasil/tree/master/S/SPIRV_LLVM_Translator uses `LLVM_full_jll` as a build dep.

Reply:
I believe `LLVM_jll` is the old builder, before Valentin and/or Elliot reorganised it. `LLVM.jl` directly calls in wathever libllvm is found loaded at runtime (hopefully Julia's one). My feeling is that we should follow `SPIRV_LLVM_Translator`'s example. @maleadt any suggestions?

Reply:
IIRC, it's been a while: I think you'll need to wait until Julia exports a `libLLVM_jll` pointing to its internal one, you can't use the current JLL because that will result in another copy of `libLLVM` getting loaded which breaks everything. And using a differently-versioned LLVM doesn't work at all. For `SPIRV_LLVM_Translator`, we use `LLVM_full` to build against a specific LLVM version, don't depend on anything at run-time because we'll pick up Julia's `libLLVM` anyway (which will have been `dlopen`ed already, unless the user has a Julia build with LLVM statically linked), and manually patch up General to ensure a compatible Translator JLL is installed: https://github.com/JuliaRegistries/General/pull/13412 https://github.com/JuliaRegistries/General/pull/14792

Reply:
Seems like just building against LLVM as per @giordano's original comment is the way to go then. I shouldn't have reverted that commit so quickly!